src/common/ui/codemirror-js-mixed/javascript-mixed.js

```diff
@@ -0,0 +1,361 @@
+        // string template to obtain tokenQuasi tokenizer
+        const dummyStream = new stream.constructor('`#dummy', 2, {});
+        const dummyState = jsMode.startState();
+        jsMode.token(dummyStream, dummyState);
+        tokenQuasi = dummyState.tokenize;
+        return tokenQuasi;
+      }
+
+      function _forceJsModeToQuasi(stream, jsState) {
+        jsState.tokenize = getTokenQuasi(stream);
+      }
+
+      return _forceJsModeToQuasi;
+    })();
+
+
+    function prepareReparseStringTemplateInLocalMode(modeToUse, stream, state) {
+      dbg(`Entering local ${modeToUse.name} mode...`);
+      // spit out beginning backtick as a token, and leave the rest of the text for local mode parsing
+      stream.backUp(stream.current().length - 1);
+
+      // workaround needed for 1-line string template,
+      // to ensure the ending backtick is parsed correctly.
+      forceJsModeToQuasi(stream, state.jsState);
+
+      // switch to local mode for subsequent text
+      state.localMode = modeToUse;
+      state.localState = CodeMirror.startState(state.localMode);
+    }
+
+    function exitLocalModeAndTokenEndingBacktick(stream, state) {
+      dbg('Exiting local html/css mode...');
+      // parse the ending JS string template backtick in js mode
+      return jsMode.token(stream, state.jsState);
+    }
+
+    function tokenInLocalMode(stream, state) {
+      const style = state.localMode.token(stream, state.localState);
+      dbg('  local mode token - ', stream.current(), `[${style}]`);
+      return style;
+    }
+
+    class Rule {
+      constructor(props) {
+        this.curContext = props.curContext;
+        // lambda for match condition
+        this.match = props.match;
+        this.nextContext = props.nextContext;
+        // optional, lambda for additional logic to run if matched
+        this.caseMatched = props.caseMatched;
+        // optional, lambda for additional logic to run if not matched
+        this.caseNotMatched = props.caseNotMatched;
+      }
+
+      run(ctx) {
+        const state = ctx.state;
+        if (this.match(ctx)) {
+          state.maybeLocalContext = this.nextContext;
+          if (state.maybeLocalContext == null) {
+            // local mode done, reset
+            state.localMode = null;
+            state.localState = null;
+          }
+          if (this.caseMatched) { this.caseMatched(ctx); }
+          return true;
+        } // case rule transition criteria not matched
+        if (this.caseNotMatched) {
+          this.caseNotMatched(ctx);
+        } else { // default not matched logic: reset local mode matching
+          state.maybeLocalContext = null;
+        }
+        return false;
+      }
+
+      // holds input parameters and return values for a rule execution
+      static createRunContext(stream, state, jsTokStyle) {
+        const ctx = {};
+
+        let _tokStr = null;
+        Object.defineProperties(ctx, {
```

Comment:
Why bothering with read/write access? This is used only internally, it's not a public API, so there should be no need for this.

Reply:
It's primarily meant as both a defensive measure and documentation. The use of context makes it unclear what are meant to be inputs and what are meant to be outputs.

In the equivalent logic before using context to pass around input/output. It's something like the following:
```javascript
function maybeToken(stream, state) {
  let tokStyle = STYLE_PASS;
  const tokTyp = state.jsState.lastType;
```
It's clear what are inputs and what are outputs (and enforced by the engine). 

Defining read/write access in the context is to serve the same purpose, to prevent developers accidentally misuse them in the future.

Reply:
I understand the general idea, I just don't see how it benefits us in this case. This code is small, self-contained, and is rather a finished product that hopefully won't see any changes so there's for so called defensive programming. As for the documentation, there's a better solution that doesn't incur a performance cost: TypeScript-compatible JSDoc comments.

Reply:
How about replacing it with a class? Something like the following:

```javascript
    class RunContext {
      constructor(stream, state, jsTokStyle) {
        /**
         * @readonly
         */
        this.stream = stream;

        /**
         * @readonly
         */
        this.state = state;

        /**
         * @readonly
         */
        this.jsTokStyle = jsTokStyle;

        /**
         * The output of a rule execution - the only writable property
         */
        this.tokStyle = STYLE_PASS;
      }

      get tokTyp() { return this.state.jsState.lastType; }
    }
```